StereoSGBM compute() fails with exception

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
- OpenCV => 3.2.0
- Operating System / Platform => Windows 7 64 bit, 8 GiB RAM, MSYS2 MinGW64 64 bit 
- Compiler => gcc 6.3.0

##### Detailed description

A call to StereoSGBM::compute() with the parameters shown in the code example fails with an exception:

> user@user-PC MINGW64 ~/opencv_stereo_sgbm_testcase
> $ ./opencv_stereo_sgbm_testcase.exe
> OpenCV Error: Assertion failed (s >= 0) in setSize, file C:/msys64/home/user/MINGW-packages-master/mingw-w64-opencv/src/opencv-3.2.0/modules/core/src/matrix.cpp, line 307
> 
> This application has requested the Runtime to terminate it in an unusual way.
> Please contact the application's support team for more information.
> terminate called after throwing an instance of 'cv::Exception'
>   what():  C:/msys64/home/user/MINGW-packages-master/mingw-w64-opencv/src/opencv-3.2.0/modules/core/src/matrix.cpp:307: error: (-215) s >= 0 in function setSize
> 
> 
> user@user-PC MINGW64 ~/opencv_stereo_sgbm_testcase
> $
> 

The compute() call succeeds with StereoSGBM::MODE_SGBM and StereoSGBM::MODE_SGBM_3WAY, but fails with StereoSGBM::MODE_HH.
The size of the input images affects this behaviour. It seems that compute() works fine when its memory consumption stays below 2 GiB.
Please note that all components involved are using a 64 bit address space, creating a 80000x80000 Mat<uint8_t> works without a problem and makes the process use 6.4 GB as expected.

##### Steps to reproduce

```.cpp
#include <opencv2/core/core.hpp>
#include <opencv2/calib3d/calib3d.hpp>

int main(void)
{
	cv::Mat src1, src0, disparity;

	src0.create(cv::Size(1812, 1737), CV_8U);
	src1.create(cv::Size(1812, 1737), CV_8U);

	cv::Ptr<cv::StereoSGBM> stereoSGBM = cv::StereoSGBM::create(0, 14*16, 3, 0, 10000, 72, 2, 1, 14, 1, 1);

	stereoSGBM->compute(src0, src1, disparity);

	return 0;
}
```

--- a/modules/calib3d/src/stereosgbm.cpp
+++ b/modules/calib3d/src/stereosgbm.cpp
@@ -353,7 +353,7 @@ static void computeDisparitySGBM( const Mat& img1, const Mat& img2,
 
     if( buffer.empty() || !buffer.isContinuous() ||
         buffer.cols*buffer.rows*buffer.elemSize() < totalBufSize )
-        buffer.create(1, (int)totalBufSize, CV_8U);
+        buffer.reserveBuffer(totalBufSize);
 
     // summary cost over different (nDirs) directions
     CostType* Cbuf = (CostType*)alignPtr(buffer.ptr(), ALIGN);
@@ -939,7 +939,7 @@ void getBufferPointers(Mat& buffer, int width, int width1, int D, int num_ch, in
                           16;  //to compensate for the alignPtr shifts
 
     if( buffer.empty() || !buffer.isContinuous() || buffer.cols*buffer.rows*buffer.elemSize() < totalBufSize )
-        buffer.create(1, (int)totalBufSize, CV_8U);
+        buffer.reserveBuffer(totalBufSize);
 
     // set up all the pointers:
     curCostVolumeLine  = (CostType*)alignPtr(buffer.ptr(), 16);
@@ -1615,7 +1615,7 @@ void filterSpecklesImpl(cv::Mat& img, int newVal, int maxSpeckleSize, int maxDif
     int width = img.cols, height = img.rows, npixels = width*height;
     size_t bufSize = npixels*(int)(sizeof(Point2s) + sizeof(int) + sizeof(uchar));
     if( !_buf.isContinuous() || _buf.empty() || _buf.cols*_buf.rows*_buf.elemSize() < bufSize )
-        _buf.create(1, (int)bufSize, CV_8U);
+        _buf.reserveBuffer(bufSize);
 
     uchar* buf = _buf.ptr();
     int i, j, dstep = (int)(img.step/sizeof(T));

--- a/modules/core/include/opencv2/core/mat.hpp
+++ b/modules/core/include/opencv2/core/mat.hpp
@@ -1416,6 +1416,14 @@ public:
      */
     void reserve(size_t sz);
 
+    /** @brief Reserves space for the certain number of bytes.
+
+    The method reserves space for sz bytes. If the matrix already has enough space to store sz bytes,
+    nothing happens. If matrix has to be reallocated its previous content could be lost.
+    @param sz Number of bytes.
+    */
+    void reserveBuffer(size_t sz);
+
     /** @brief Changes the number of matrix rows.
 
     The methods change the number of matrix rows. If the matrix is reallocated, the first

--- a/modules/core/src/matrix.cpp
+++ b/modules/core/src/matrix.cpp
@@ -830,6 +830,32 @@ void Mat::reserve(size_t nelems)
     dataend = data + step.p[0]*r;
 }
 
+void Mat::reserveBuffer(size_t nbytes)
+{
+    size_t esz = 1;
+    int mtype = CV_8UC1;
+    if (!empty())
+    {
+        if (!isSubmatrix() && data + nbytes <= dataend)//Should it be datalimit?
+            return;
+        esz = elemSize();
+        mtype = type();
+    }
+
+    size_t nelems = (nbytes - 1) / esz + 1;
+
+#if SIZE_MAX > UINT_MAX
+    CV_Assert(nelems <= size_t(INT_MAX)*size_t(INT_MAX));
+    int newrows = nelems > size_t(INT_MAX) ? nelems > 0x400*size_t(INT_MAX) ? nelems > 0x100000 * size_t(INT_MAX) ? nelems > 0x40000000 * size_t(INT_MAX) ?
+                  size_t(INT_MAX) : 0x40000000 : 0x100000 : 0x400 : 1;
+#else
+    int newrows = nelems > size_t(INT_MAX) ? 2 : 1;
+#endif
+    int newcols = (int)((nelems - 1) / newrows + 1);
+
+    create(newrows, newcols, mtype);
+}
+
 
 void Mat::resize(size_t nelems)
 {

